### ISSUE TITLE
ModuleNotFoundError returned while connecting via ssh: No module named 'invoke'

### ISSUE THREAD

**Issue by andrewtdavis** [Juniper/splitcopy#4]
When attempting to run splitcopy in a Python3 virtual environment, the error `ModuleNotFoundError returned while connecting via ssh: No module named 'invoke'` is returned. Manually installing invoke fixed it, but it doesn't look like pip knows it's required.

This is on macOS Monterey.

Steps to reproduce:
1. `python3 -m venv juniper`
2. `source juniper/bin/env`
3. `pip install splitcopy` (I Had errors the first time with `cryptography`, not sure if it was me though because I forgot to upgrade pip first.)
4. `pip install splitcopy` (again, because mine failed the first time)
5. `splitcopy Downloads/junos-srxsme-20.2R3-S2.5.tgz user@192.168.1.1:/var/tmp`

**Comment by ipmonk** [comment]
Hi Andrew, which python3/paramiko version are you using? which pip version were you using initially? 
It'd be good to be able to reproduce this. 

Splitcopy does not directly import 'invoke'. 
The only 3rd party module that splitcopy imports that does import 'invoke' is paramiko. 
pip/pypi should track these dependancies and install them as needed.

**Comment by andrewtdavis** [comment]
I re-built my venv to re-test, this time upgrading pip first and can confirm that did fix the `cryptography` error. So the only error I have now is invoke, but it did happen again:

PIP: `pip 20.2.3 from .../juniper/lib/python3.8/site-packages/pip (python 3.8)`
Post-upgrade: `pip 22.1.2 from juniper/lib/python3.8/site-packages/pip (python 3.8)`

Full install after pip upgrade:
```
(juniper) user@host ~ % pip install splitcopy
Collecting splitcopy
  Downloading splitcopy-1.4.0-py3-none-any.whl (36 kB)
Collecting scp
  Downloading scp-0.14.4-py2.py3-none-any.whl (8.6 kB)
Collecting paramiko
  Downloading paramiko-2.11.0-py2.py3-none-any.whl (212 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.9/212.9 kB 3.8 MB/s eta 0:00:00
Collecting pynacl>=1.0.1
  Downloading PyNaCl-1.5.0-cp36-abi3-macosx_10_10_universal2.whl (349 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 349.9/349.9 kB 6.0 MB/s eta 0:00:00
Collecting cryptography>=2.5
  Downloading cryptography-37.0.2-cp36-abi3-macosx_10_10_x86_64.whl (2.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.8/2.8 MB 15.1 MB/s eta 0:00:00
Collecting six
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting bcrypt>=3.1.3
  Downloading bcrypt-3.2.2-cp36-abi3-macosx_10_10_universal2.whl (50 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.0/50.0 kB 1.8 MB/s eta 0:00:00
Collecting cffi>=1.1
  Downloading cffi-1.15.0-cp38-cp38-macosx_10_9_x86_64.whl (178 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 178.7/178.7 kB 6.4 MB/s eta 0:00:00
Collecting pycparser
  Downloading pycparser-2.21-py2.py3-none-any.whl (118 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.7/118.7 kB 4.9 MB/s eta 0:00:00
Installing collected packages: six, pycparser, cffi, pynacl, cryptography, bcrypt, paramiko, scp, splitcopy
Successfully installed bcrypt-3.2.2 cffi-1.15.0 cryptography-37.0.2 paramiko-2.11.0 pycparser-2.21 pynacl-1.5.0 scp-0.14.4 six-1.16.0 splitcopy-1.4.0
```

Attempting to run splitcopy gave the same `ModuleNotFoundError returned while connecting via ssh: No module named 'invoke'` error.

**Comment by ipmonk** [comment]
Hi Andrew, 

I've tried many different ways to reproduce this and so far have been unable to (older python versions etc)
Most recently I used a macOS VM which is as close to your setup as I can get. 
Please see below for the test results:

// using the default version of python shipped with MacOS, matching the pip version, splitcopy version and all the dependencies versions above:

foobars-Virtual-Machine:~ foobar$ sw_vers
ProductName:	macOS
ProductVersion:	12.5
BuildVersion:	21G72
foobars-Virtual-Machine:~ foobar$ python3 -V
Python 3.8.9
foobars-Virtual-Machine:~ foobar$ python3 -m venv juniper
foobars-Virtual-Machine:~ foobar$
foobars-Virtual-Machine:~ foobar$
foobars-Virtual-Machine:~ foobar$ cd juniper
foobars-Virtual-Machine:juniper foobar$ source bin/activate
(juniper) foobars-Virtual-Machine:juniper foobar$ python3 -m pip install -U pip==22.1.2
Collecting pip==22.1.2
  Using cached pip-22.1.2-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 20.2.3
    Uninstalling pip-20.2.3:
      Successfully uninstalled pip-20.2.3
Successfully installed pip-22.1.2
(juniper) foobars-Virtual-Machine:juniper foobar$ python3 -m pip install bcrypt==3.2.2 cffi==1.15.0 cryptography==37.0.2 paramiko==2.11.0 pycparser==2.21 pynacl==1.5.0 scp==0.14.4 six==1.16.0
Collecting bcrypt==3.2.2
  Using cached bcrypt-3.2.2-cp36-abi3-macosx_10_10_universal2.whl (50 kB)
Collecting cffi==1.15.0
  Downloading cffi-1.15.0.tar.gz (484 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 484.1/484.1 kB 3.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Collecting cryptography==37.0.2
  Downloading cryptography-37.0.2-cp36-abi3-macosx_10_10_universal2.whl (5.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.3/5.3 MB 10.2 MB/s eta 0:00:00
Collecting paramiko==2.11.0
  Using cached paramiko-2.11.0-py2.py3-none-any.whl (212 kB)
Collecting pycparser==2.21
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Collecting pynacl==1.5.0
  Using cached PyNaCl-1.5.0-cp36-abi3-macosx_10_10_universal2.whl (349 kB)
Collecting scp==0.14.4
  Using cached scp-0.14.4-py2.py3-none-any.whl (8.6 kB)
Collecting six==1.16.0
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Using legacy 'setup.py install' for cffi, since package 'wheel' is not installed.
Installing collected packages: six, pycparser, cffi, pynacl, cryptography, bcrypt, paramiko, scp
  Running setup.py install for cffi ... done
Successfully installed bcrypt-3.2.2 cffi-1.15.0 cryptography-37.0.2 paramiko-2.11.0 pycparser-2.21 pynacl-1.5.0 scp-0.14.4 six-1.16.0

[notice] A new release of pip available: 22.1.2 -> 22.2.2
[notice] To update, run: pip install --upgrade pip
(juniper) foobars-Virtual-Machine:juniper foobar$ python3 -m pip install splitcopy==1.4.0
Collecting splitcopy==1.4.0
  Using cached splitcopy-1.4.0-py3-none-any.whl (36 kB)
Requirement already satisfied: scp in ./lib/python3.8/site-packages (from splitcopy==1.4.0) (0.14.4)
Requirement already satisfied: paramiko in ./lib/python3.8/site-packages (from splitcopy==1.4.0) (2.11.0)
Requirement already satisfied: pynacl>=1.0.1 in ./lib/python3.8/site-packages (from paramiko->splitcopy==1.4.0) (1.5.0)
Requirement already satisfied: six in ./lib/python3.8/site-packages (from paramiko->splitcopy==1.4.0) (1.16.0)
Requirement already satisfied: cryptography>=2.5 in ./lib/python3.8/site-packages (from paramiko->splitcopy==1.4.0) (37.0.2)
Requirement already satisfied: bcrypt>=3.1.3 in ./lib/python3.8/site-packages (from paramiko->splitcopy==1.4.0) (3.2.2)
Requirement already satisfied: cffi>=1.1 in ./lib/python3.8/site-packages (from bcrypt>=3.1.3->paramiko->splitcopy==1.4.0) (1.15.0)
Requirement already satisfied: pycparser in ./lib/python3.8/site-packages (from cffi>=1.1->bcrypt>=3.1.3->paramiko->splitcopy==1.4.0) (2.21)
Installing collected packages: splitcopy
Successfully installed splitcopy-1.4.0

[notice] A new release of pip available: 22.1.2 -> 22.2.2
[notice] To update, run: pip install --upgrade pip

(juniper) foobars-Virtual-Machine:juniper foobar$ splitcopy v6_linklocal_isis_db root@10.85.174.69:
root@10.85.174.69's password:
ssh authentication succeeded
attempting FTP authentication...
checking remote storage...
checking local storage...
sha1 not found, generating sha1...
splitting file...
starting transfer...

transfer complete
joining chunks...
deleting remote tmp directory...
generating remote sha hash...
local and remote sha hash match
file has been successfully copied to 10.85.174.69:~/v6_linklocal_isis_db
data transfer = 0:00:03.026563
total runtime = 0:00:21.144840

// the only difference I can see is step 2 in your original post, as juniper/bin/env doesn't exist in my setup?

I'm loathe to add a dependency to this project, when the problem relates to pip somehow.

**Comment by ipmonk** [comment]
closing as unreproducible
